antenna/configlib.py

```diff
@@ -0,0 +1,323 @@
+    the ``os.environ``.
+
+    """
+    def get(self, key, namespace=''):
+        if namespace:
+            key = '%s_%s' % (namespace, key)
+
+        key = key.upper()
+        if key in os.environ:
+            return os.environ[key]
+
+        return NO_VALUE
+
+
+class ConfigIniEnv(object):
+    """.ini style configuration layer
+
+    Namespace is a config section. So "foo" in namespace "bar" is::
+
+        [bar]
+        foo=someval
+
+    """
+    def __init__(self, fn):
+        self._parser = ConfigParser()
+        self._parser.readfp(open(fn, 'r'))
+
+    def get(self, key, namespace='main'):
+        if self._parser.has_option(namespace, key):
+            return self._parser.get(namespace, key)
+        return NO_VALUE
+
+
+class ConfigManager(object):
+    """Manages multiple configuration environment layers"""
+    def __init__(self):
+        self.envs = self._initialize_environments()
+
+    def _initialize_environments(self):
+        # Add environments in order of precedence--first is the most important!
+        envs = []
+
+        # Add override environment
+        envs.append(ConfigOverrideEnv())
+
+        # Add OS environment
+        envs.append(ConfigOSEnv())
+
+        # Add .ini environment
+        fn = os.environ.get('ANTENNA_INI', 'antenna.ini')
+        if not os.sep in fn:
+            fn = os.path.join(os.getcwd(), fn)
+        if os.path.exists(fn):
+            envs.append(ConfigIniEnv(fn))
+
+        return envs
+
+    def __call__(self, key, namespace=None, default=NO_VALUE, parser=str,
+                 raise_error=False):
+        """Returns a parsed value from the environment
+
+        :arg key: the key to look up
+        :arg namespace: the namespace for the key--different environments
+            use this differently
+        :arg default: the default value (if any)
+        :arg parser: the parser for converting this value to a Python object
+        :arg raise_error: True if you want a lack of value to raise a
+            ``ConfigurationError``
+
+        Examples::
+
+            # Use the special bool parser
+            DEBUG = config('DEBUG', default=True, parser=bool)
+
+            from antenna.config_util import ListOf
+            ALLOWED_HOSTS = config('ALLOWED_HOSTS', default='localhost',
+                                   parser=ListOf(str))
+
+        """
+        parser = get_parser(parser)
+
+        # Go through environments in reverse order
+        for env in self.envs:
+            val = env.get(key, namespace)
+            if val is not NO_VALUE:
+                return parser(val)
+
+        # Return the default if there is one
+        if default is not NO_VALUE:
+            return parser(default)
+
+        # No value specified and no default, so raise an error to the user
+        if raise_error:
+            raise ConfigurationError(
+                '%s (%s) requires a value of type %s' % (key, namespace, parser)
+            )
```

Comment:
I don't know what you mean. Can you show an example of the typo you're talking about?